src/main/java/htsjdk/samtools/read/Read.java

```diff
@@ -0,0 +1,1438 @@
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+package htsjdk.samtools.read;
+
+import htsjdk.samtools.Cigar;
+import htsjdk.samtools.SAMException;
+import htsjdk.samtools.SAMFlag;
+import htsjdk.samtools.SAMUtils;
+import htsjdk.samtools.util.Locatable;
+import htsjdk.samtools.util.StringUtil;
+
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+
+/**
+ * Interface for read, which is a sequence of bases aligned to a reference.
+ *
+ * <p>This interface is based on the <a href="http://samtools.github.io/hts-specs/SAMv1.pdf">SAM specifications</a>.
+ *
+ * @author Daniel Gomez-Sanchez (magicDGS)
+ */
+public interface Read extends Locatable {
+
+    ///////////////////////////////////////////////////////////
+    // MANDATORY FIELDS - getters/setters
+    ///////////////////////////////////////////////////////////
+
+    /**
+     * Gets the name of the read.
+     *
+     * <p>Equivalent to <b>QNAME</b> in the <a href="http://samtools.github.io/hts-specs/SAMv1.pdf">SAM specifications</a>.
+     *
+     * @return the read name or {@code null} if the read has no name.
+     */
+    String getName();
+
+    /**
+     * Sets the name of the read.
+     *
+     * <p>Equivalent to <b>QNAME</b> in the <a href="http://samtools.github.io/hts-specs/SAMv1.pdf">SAM specifications</a>.
+     *
+     * @param name new name for the read; {@code null} if the read has no name.
+     */
+    void setName(final String name);
+
+    /**
+     * Gets the bitwise flag.
+     *
+     * <p>Equivalent to <b>FLAG</b> in the <a href="http://samtools.github.io/hts-specs/SAMv1.pdf">SAM specifications</a>.
+     *
+     * <p>Note: it is preferable to use the methods for testing if specific flags are set (using {@link #isSet(SAMFlag)} or {@link #isUnset(SAMFlag)}),
+     * get all the set flags with {@link #getSetFlags()} or testing specific flags with the shortcuts.
+     *
+     * @return the bitwise flag as an integer.
+     */
+    int getFlag();
+
+    /**
+     * Gets the {@link SAMFlag} that are set.
+     *
+     * <p>Equivalent to <b>FLAG</b> in the <a href="http://samtools.github.io/hts-specs/SAMv1.pdf">SAM specifications</a>.
+     *
+     * <p>Default implementation calls {@code SAMFlag.getFlags(getFlag())}.
+     *
+     * @return a set of the flags that are set for the read.
+     */
+    default Set<SAMFlag> getSetFlags() {
```

Comment:
this is a rather unfortunate name..."set" as an adjective....rename perhaps?

Reply:
Thanks! I know that there are going to be a lot of discussion about naming, and that's why I submitted the PR. Any suggestion for a new name?

Reply:
actually, it's not clear if you meant "set" as a noun perhaps.....Argh. I don't have a better suggestion though.

Reply:
how about `getTrueFlags`?

Reply:
I think that after you pointed out the language problem with the flags, we should really think carefully about the naming of all the flag methods:

* Setters: we should distinguish between single/all bits, as you pointed out below. `Flag` vs `Flags` is a good suggestion, but maybe we can be more clear/explicit. Because using the int flag is kind of discourage, what's about calling the ones setting all the bits `setRawFlag()` and `getRawFlag()`. Or maybe `Bitwise` instead of `Raw`.
* Getters: for single values, we have the `isSet` and `isUnset`. The problem comes with methods like the one pointed here: get all the flags that are set or unset. If we change the nomenclature to `getTrueFlags()` and `getFalseFlags()`, they would be out-of-sync with other getters/setters.

This name is kind of difficult to solve. Any proposal for all the methods will be welcomed!